relay/relay.go

```diff
@@ -255,46 +275,64 @@ func (rs *Relay) GetHeader(ctx context.Context, m *structs.MetricGroup, uc struc
 	m.AppendSince(tGet, "getHeader", "get")
 
 	key := structs.PayloadKey{
-		BlockHash: maxProfitBlock.Header.Trace.BlockHash,
-		Slot:      structs.Slot(maxProfitBlock.Header.Trace.Slot),
-		Proposer:  maxProfitBlock.Header.Trace.ProposerPubkey}
-	if err := rs.d.CacheBlock(ctx, key, maxProfitBlock); err != nil {
-		logger.Warnf("fail to cache block: %s", err.Error())
-	}
-	logger.Debug("payload cached")
+		// TODO: BlockHash: maxProfit.BuilderBid.,
+		Slot:     structs.Slot(maxProfit.Slot),
+		Proposer: maxProfit.Proposer}
 
-	header := maxProfitBlock.Header
-	if header.Header == nil {
+	header := maxProfit.BuilderBid.Header()
+	if header == nil {
 		rs.m.MissHeaderCount.WithLabelValues("badHeader").Add(1)
 		return nil, ErrNoBuilderBid
 	}
 
-	if header.Header.GetParentHash() != parentHash {
-		logger.WithField("expected", header.Header.GetParentHash()).WithField("got", parentHash).Debug("invalid parentHash")
+	if header.GetParentHash() != parentHash {
+		logger.WithField("expected", header.GetParentHash()).WithField("got", parentHash).Debug("invalid parentHash")
 		rs.m.MissHeaderCount.WithLabelValues("badHeader").Add(1)
 		return nil, ErrNoBuilderBid
 	}
 
-	if header.Trace.ProposerPubkey != pk.PublicKey {
-		logger.WithField("expected", header.Trace.BuilderPubkey).WithField("got", pk.PublicKey).Debug("invalid pubkey")
+	if maxProfit.Proposer != pk.PublicKey {
+		logger.WithField("expected", maxProfit.Proposer).WithField("got", pk.PublicKey).Debug("invalid pubkey")
 		rs.m.MissHeaderCount.WithLabelValues("badHeader").Add(1)
 		return nil, ErrNoBuilderBid
 	}
 
-	if zero := types.IntToU256(0); header.Trace.Value.Cmp(&zero) == 0 {
+	value := maxProfit.BuilderBid.Value()
+	if zero := types.IntToU256(0); value.Cmp(&zero) == 0 {
 		rs.m.MissHeaderCount.WithLabelValues("zeroBid").Add(1)
 		return nil, ErrZeroBid
 	}
 
 	fork := rs.beaconState.ForkVersion(slot)
+	go func() {
```

Comment:
I'm not sure can we do that in goroutine. as it has to be stored (and probably returned) prior to  returning that to the user. Otherwise there can be a operation where we would not have data stored on getpayload

Reply:
+ if we do that we need not to fail on "return"  from GetPayload. Also I would call PublishBlockCache as soon as possible.